src/package-tree.ts

```diff
@@ -106,25 +111,29 @@ function resolvePaths(
 
 /**
  * Recursive function for resolvePaths
- * 
+ *
  * @param packageNode the original package, where the data property is null
  * @param parentPath The parent path of this current packageTree node
  */
-function resolvePathsRec(
-    packageNode: PackageTree<null>, parentPath: string): PackageTree<string> {
+async function resolvePathsRec(
+    packageNode: PackageTree<null>,
+    parentPath: string): Promise<PackageTree<string>> {
   const paths: string[] = [];
   const resolvedNodes: Array<PackageTree<string>> = [];
   paths.push(parentPath);
-  const path = require.resolve(packageNode.name, {paths});
+  let currPath = path.dirname(require.resolve(packageNode.name, {paths}));
+  while (!(await filesInDir(currPath)).includes('package.json')) {
+    currPath = path.dirname(require.resolve(packageNode.name, {paths}));
+  }
 
-  packageNode.dependencies.forEach((child) => {
-    resolvedNodes.push(resolvePathsRec(child, parentPath));
-  });
+  await Promise.all(packageNode.dependencies.map(async (child) => {
+    resolvedNodes.push(await resolvePathsRec(child, currPath));
+  }));
 
   const updatedNode: PackageTree<string> = {
     name: packageNode.name,
     version: packageNode.version,
-    data: path,
+    data: currPath,
```

Comment:
I think my question still stands -- if `data` already contains a value, is it safe to override it?

Reply:
yes, if a value already exists in data, it means that there are more than 1 package that depends on the same version of this package. Every package should have a unique path, so the new data value and the existing data value should be the same.

Reply:
As you stated, there is an invariant that if `data` already exists, then the new value `currPath` must have the same value. Let's throw an error if this invariant is not satisfied.